src/main/java/org/embulk/input/zendesk/services/ZendeskSupportAPIService.java

```diff
@@ -0,0 +1,145 @@
+package org.embulk.input.zendesk.services;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import org.embulk.input.zendesk.ZendeskInputPlugin.PluginTask;
+import org.embulk.input.zendesk.clients.ZendeskRestClient;
+import org.embulk.input.zendesk.models.Target;
+import org.embulk.input.zendesk.utils.ZendeskConstants;
+import org.embulk.input.zendesk.utils.ZendeskUtils;
+import org.embulk.spi.DataException;
+
+import java.io.IOException;
+
+public class ZendeskSupportAPIService
+{
+    private ZendeskRestClient zendeskRestClient;
+
+    private PluginTask task;
+
+    private static ObjectMapper mapper = new ObjectMapper();
+
+    static {
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        mapper.configure(com.fasterxml.jackson.core.JsonParser.Feature.ALLOW_UNQUOTED_CONTROL_CHARS, false);
+    }
+
+    public ZendeskSupportAPIService(final PluginTask task)
+    {
+        this.task = task;
+    }
+
+    public void setTask(PluginTask task)
+    {
+        this.task = task;
+    }
+
+    public JsonNode getData(String path, final int page, final boolean isPreview, long startTime)
+    {
+        if (path.isEmpty()) {
+            path = buildPath(page, isPreview, startTime);
+        }
+        try {
+            final String response = getZendeskRestClient().doGet(path, task);
+            return parseJsonObject(response);
+        }
+        catch (final Exception e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private ObjectNode parseJsonObject(final String jsonText)
+    {
+        JsonNode node = parseJsonNode(jsonText);
+        if (node.isObject()) {
+            return (ObjectNode) node;
+        }
+        else {
+            throw new DataException("Expected object node: " + jsonText);
+        }
+    }
+
+    private JsonNode parseJsonNode(final String jsonText)
+    {
+        try {
+            return mapper.readTree(jsonText);
+        }
+        catch (final IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private String buildPath(final int page, final boolean isPreview, long startTime)
+    {
+        final StringBuilder path = isPreview
+                ? new StringBuilder(buildURLForPreview())
+                : new StringBuilder(buildURLForRun(page, startTime));
+
+        if (Target.TICKET_METRICS.equals(task.getTarget())) {
+            path.append("&include=metric_sets");
+        }
+
+        return path.toString();
+    }
+
+    private String buildURLForPreview()
```

Comment:
I saw 90% code snippet between `buildURLForPreview` and `buildURLForRunBasedOnTarget`, is it possible to parameterize the different then build 1 method only?

Reply:
fixed